site-graph/http/http.go

```diff
@@ -0,0 +1,60 @@
+package http
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/adamplansky/go-bridge-mentoring/site-graph/cache"
+
+	"github.com/gorilla/mux"
+	"go.uber.org/zap"
+
+	"github.com/adamplansky/go-bridge-mentoring/site-graph/crawler"
+)
+
+type server struct {
+	router  *mux.Router
+	log     *zap.SugaredLogger
+	crawler *crawler.Crawler
+	cache   cache.Cache
+}
+
+func NewServer(log *zap.SugaredLogger, c cache.Cache) *server {
+	s := server{
+		router:  mux.NewRouter(),
+		log:     log,
+		crawler: crawler.New(log, c),
+		cache:   c,
+	}
+	s.routes()
+	return &s
+}
+
+func (s *server) Run(ctx context.Context, addr string) error {
+	httpServer := http.Server{
+		Addr:           addr,
+		Handler:        s.router,
+		ReadTimeout:    15 * time.Second,           // the maximum duration for reading the entire request, including the body
+		WriteTimeout:   15 * time.Second,           // the maximum duration before timing out writes of the response
+		IdleTimeout:    30 * time.Second,           // the maximum amount of time to wait for the next request when keep-alive is enabled
+		MaxHeaderBytes: http.DefaultMaxHeaderBytes, // 1 MB
+		BaseContext: func(_ net.Listener) context.Context {
```

Comment:
this will make request context cancel on server shutdown - so in handler it will be not visible - is it user closing connection of server trying to shutdown. 
if you want real graceful shutdown - handler should execute for the time when user is making it 
Only if handler is working too long it will be interrupted by shutdownCtx bellow